controllers/toolchaincluster_resources/toolchaincluster_resources_controller.go

```diff
@@ -0,0 +1,68 @@
+package toolchaincluster_resources
+
+import (
+	"context"
+	"embed"
+	"fmt"
+
+	toolchainv1alpha1 "github.com/codeready-toolchain/api/api/v1alpha1"
+	commoncontroller "github.com/codeready-toolchain/toolchain-common/controllers"
+	applycl "github.com/codeready-toolchain/toolchain-common/pkg/client"
+	commonpredicates "github.com/codeready-toolchain/toolchain-common/pkg/predicate"
+	"github.com/codeready-toolchain/toolchain-common/pkg/template"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	runtimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *Reconciler) SetupWithManager(mgr ctrl.Manager, operatorNamespace string) error {
+	build := ctrl.NewControllerManagedBy(mgr).
+		For(&v1.ServiceAccount{})
+
+	// add watcher for all kinds from given templates
+	allObjects, err := template.LoadObjectsFromEmbedFS(r.templates, &template.Variables{Namespace: operatorNamespace})
+	if err != nil {
+		return err
+	}
+	mapToOwnerByLabel := handler.EnqueueRequestsFromMapFunc(commoncontroller.MapToOwnerByLabel("", toolchainv1alpha1.ProviderLabelKey))
+	for _, obj := range allObjects {
+		build = build.Watches(&source.Kind{Type: obj.DeepCopyObject().(runtimeclient.Object)}, mapToOwnerByLabel, builder.WithPredicates(commonpredicates.LabelsAndGenerationPredicate{}))
+	}
+	return build.Complete(r)
+}
+
+// Reconciler reconciles a ToolchainCluster object
+type Reconciler struct {
+	client    runtimeclient.Client
+	scheme    *runtime.Scheme
+	templates *embed.FS
+}
+
+// Reconcile loads all the manifests from a given embed.FS folder, evaluates the supported variables and applies the objects in the cluster.
+func (r *Reconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	reqLogger := log.FromContext(ctx)
+	reqLogger.Info("Reconciling ToolchainCluster resources controller")
+	// check for required templates FS directory
+	if r.templates == nil {
+		return reconcile.Result{}, fmt.Errorf("no templates FS configured")
+	}
+
+	// read the template fs
+	allObjects, err := template.LoadObjectsFromEmbedFS(r.templates, &template.Variables{Namespace: request.Namespace})
```

Comment:
btw, you already read, parse, and process the template when you are initializing the controller, we pass only one parameter (namespace) that won't change during the execution, so you could store the retrieved objects in the `Reconciler` type and avoid doing the same template process over and over again.
WDYT?

Reply:
that's a good point. It will avoid to reload the templates every time ( even if this is done in memory , so it will be very quick ). 
The only "down side" of this is that we don't actually unit test the `SetupWithManager` function right ? So the loading part will be uncovered. But I think that's fine. 

WDYT?

Reply:
yeah, you already test the `template.LoadObjectsFromEmbedFS` function in a unit-test, so we should be fine :+1:

Reply:
I've moved the initialization of the templates in https://github.com/codeready-toolchain/toolchain-common/pull/373/commits/fda181c169cfeb20493405a036d363a5dbac558b 

Thanks!

Reply:
ok, I just submitted a review for the original code - before you pushed the commit :man_facepalming:

Reply:
yeah, sorry about that! I wanted to make sure we agree on this comment before updating the code 😬

Reply:
just go through my comments - you already addressed some of them in your commit